src/categorical_algebra/CSets.jl

```diff
@@ -1429,6 +1433,271 @@ end
     end...))
 end
 
+
+"""A map f (from A to B) as a map of subobjects of A to subjects of B"""
+(f::ACSetTransformation)(X::SubACSet) = begin
+  codom(hom(X)) == dom(f) || error("Cannot apply $f to $X")
+  Subobject(codom(f); Dict(map(ob(acset_schema(dom(f)))) do o
+    o => sort(unique(f[o].(collect(components(X)[o]))))
+  end)...)
+end
+
+
+"""
+A map f (from A to B) as a map from A to a subobject of B
+# i.e. get the image of f as a subobject of B
+"""
+(f::ACSetTransformation)(X::StructACSet) =
+  X == dom(f) ? f(top(X)) : error("Cannot apply $f to $X")
+
+"""    preimage(f::ACSetTransformation,Y::Subobject)
+Inverse of f (from A to B) as a map of subobjects of B to subjects of A.
+"""
+function preimage(f::ACSetTransformation,Y::SubACSet)
+  codom(hom(Y)) == codom(f) || error("Cannot apply $f to $X")
+  Subobject(dom(f); Dict{Symbol, Vector{Int}}(map(ob(acset_schema(dom(f)))) do o
+    o => sort(unique(vcat([preimage(f[o],y) for y in collect(components(Y)[o])]...)))
+  end)...)
+end
+
+"""    preimage(f::CSetTransformation,Y::StructACSet)
+Inverse f (from A to B) as a map from subobjects of B to subobjects of A.
+Cast an ACSet to subobject, though this has a trivial answer when computing
+the preimage (it is necessarily the top subobject of A).
+"""
+preimage(f::CSetTransformation,Y::StructACSet) =
+  Y == codom(f) ? top(dom(f)) : error("Cannot apply inverse of $f to $Y")
+
+# VarACSets
+###########
+
+"""
+For any ACSet, X, a canonical map A→X where A has distinct variables for all
+subparts.
+"""
+function abstract(X::ACSet)
```

Comment:
I'm actually surprised that `abstract` isn't a reserved word in Julia (I guess only `abstract type` is) but in any case this name seems a bit too generic. Maybe `abstract_attributes`?

Reply:
Resolved in latest commit.